lib/grp.gi

```diff
@@ -130,14 +130,103 @@ end);
 InstallOtherMethod(MinimalGeneratingSet,"fallback method to inform user",true,
   [IsObject],0,
 function(G)
+  local
+    check,GbyGk,
+    Gkm1byGk,
+    Gkm1byGk_L,
+    Gkm1byGk_elem_reps,
+    Gkm1byGk_gen,
+    Gkm1byGk_gen_reps,
+    mingenset_km1_reps,
+    mingenset_k_reps,
+    temp,i,j,l,L,x,xl,y,prev,gmod,N,g,g0,g1,s,r,stop,
+    cs,phi_GbyG1,GbyG1,phi_GbyGk,phi_Gkm1byGk,k;
   if IsGroup(G) and IsSolvableGroup(G) then
     TryNextMethod();
+  elif IsGroup(G) and IsFinite(G) then
+    cs := ChiefSeries(G);
+    phi_GbyG1 := NaturalHomomorphismByNormalSubgroup(G,cs[2]);
+    GbyG1 := ImagesSource(phi_GbyG1);
+    mingenset_k_reps := List(MinimalGeneratingSet(GbyG1), x -> PreImagesRepresentative(phi_GbyG1, x));
+    # GbyG1 is a simple group, so it can be handelled by GbyG1
+    for k in [3..Length(cs)] do # Lifting
+      mingenset_km1_reps := mingenset_k_reps;
+      phi_GbyGk := NaturalHomomorphismByNormalSubgroup(G,cs[k]);
+      phi_Gkm1byGk := NaturalHomomorphismByNormalSubgroup(cs[k-1],cs[k]);
+      Gkm1byGk := ImagesSource(phi_Gkm1byGk);
+      GbyGk := ImagesSource(phi_GbyGk);
+      check := gx -> GbyGk = GroupByGenerators(ImagesSet(phi_GbyGk,gx));
+      Gkm1byGk_elem_reps := List(AsList(Gkm1byGk),x -> PreImagesRepresentative(phi_Gkm1byGk,x));
+      Gkm1byGk_gen := SmallGeneratingSet(Gkm1byGk);
+      Gkm1byGk_gen_reps := List(Gkm1byGk_gen,x -> PreImagesRepresentative(phi_Gkm1byGk,x));
+      g := ShallowCopy(mingenset_km1_reps);
+      stop := false;
+      if IsAbelian(Gkm1byGk) then
+        if check(g) then mingenset_k_reps := g; fi;
+        for i in [1..Length(g)] do
+          if stop then break; fi;
+          for j in [1..Length(Gkm1byGk_gen_reps)] do
+            temp := g[i];
+            g[i] := temp * Gkm1byGk_gen_reps[j];
+            if check(g) then
+              mingenset_k_reps := g;
+              stop := true;
+              break;
+            fi;
+            g[i] := temp;
+          od;
+        od;
+        if not stop then
+          Add(g,Gkm1byGk_gen_reps[1]);
+          if check(g) then
+            mingenset_k_reps := g;
+          else
+            Error("The algorithm is failing");
+          fi;
+        fi;
+      else
+          g0 := ShallowCopy(mingenset_km1_reps);
+          g := ShallowCopy(mingenset_km1_reps);
+          Add(g,Gkm1byGk_elem_reps[1]);
+          g1 := ShallowCopy(g);
+          for g in [g0,g1] do
```

Comment:
So in line 192 you overwrite `g` anyway. So why is `g1` yet another copy of `g`? Why not like this?
```suggestion
          g1 := ShallowCopy(mingenset_km1_reps);
          Add(g1,Gkm1byGk_elem_reps[1]);
          for g in [g0,g1] do
```

Reply:
Fixed.